File: frontend/mobile/apple/iOS/My Home Agent/My Home Agent/ViewModels/LoginViewModel.swift
import Foundation
import SwiftUI

@MainActor
class LoginViewModel: ObservableObject {
    @Published var email = ""
    @Published var password = ""
    @Published var rememberMe = false
    @Published var isLoading = false
    @Published var error: Error?
    @Published var isAuthenticated = false
    @Published var canUseBiometrics = false
    @Published var showErrorAlert = false
    
    func login() async {
        print("📝 LoginViewModel: Starting login process")
        print("📝 LoginViewModel: Email: \(email), RememberMe: \(rememberMe)")
        isLoading = true
        error = nil
        
        do {
            print("📝 LoginViewModel: Calling AuthManager.login")
            _ = try await AuthManager.shared.login(email: email, password: password, rememberMe: rememberMe)
            print("📝 LoginViewModel: Login successful")
            isAuthenticated = true
            isLoading = false
        } catch {
            print("❌ LoginViewModel: Login failed with error: \(error)")
            print("❌ LoginViewModel: Error description: \(error.localizedDescription)")
            
            // Store the error for display
            self.error = error
            showErrorAlert = true
            isLoading = false
        }
    }
    
    func loginWithBiometrics() async {
        guard AuthManager.shared.canUseBiometricAuthentication() else {
            error = AuthError.biometricError("Biometric authentication not available")
            showErrorAlert = true
            return
        }
        
        isLoading = true
        error = nil
        
        do {
            _ = try await AuthManager.shared.authenticateWithBiometrics()
            isAuthenticated = true
            isLoading = false
        } catch {
            // Store the error for display
            self.error = error
            showErrorAlert = true
            isLoading = false
        }
    }
    
    func logout() {
        AuthManager.shared.logout()
        isAuthenticated = false
    }
    
    // Check if user is already authenticated
    func checkAuthentication() {
        isAuthenticated = AuthManager.shared.isAuthenticated
        
        // Auto-fill email if stored
        if let storedEmail = AuthManager.shared.storedEmail {
            email = storedEmail
        }
        
        // Set remember me state
        rememberMe = AuthManager.shared.isRememberMeEnabled
        
        // Check if biometric authentication is available
        canUseBiometrics = AuthManager.shared.canUseBiometricAuthentication()
    }
}

File: frontend/mobile/google/android/MyHomeAgent/app/src/main/java/com/hestami_ai/myhomeagent/ui/viewmodel/LoginViewModel.kt
package com.hestami_ai.myhomeagent.ui.viewmodel

import android.app.Application
import android.util.Log
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.hestami_ai.myhomeagent.auth.AuthError
import com.hestami_ai.myhomeagent.auth.AuthManager
import com.hestami_ai.myhomeagent.data.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Login screen UI state.
 */
data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val rememberMe: Boolean = false,
    val isLoading: Boolean = false,
    val error: String? = null,
    val isAuthenticated: Boolean = false,
    val canUseBiometrics: Boolean = false,
    val user: User? = null
)

/**
 * ViewModel for the login screen.
 * Matches iOS LoginViewModel.swift implementation.
 */
class LoginViewModel(application: Application) : AndroidViewModel(application) {

    companion object {
        private const val TAG = "LoginViewModel"
    }

    private val authManager = AuthManager.getInstance(application)

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    init {
        checkAuthentication()
    }

    fun updateEmail(email: String) {
        _uiState.value = _uiState.value.copy(email = email, error = null)
    }

    fun updatePassword(password: String) {
        _uiState.value = _uiState.value.copy(password = password, error = null)
    }

    fun updateRememberMe(rememberMe: Boolean) {
        _uiState.value = _uiState.value.copy(rememberMe = rememberMe)
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    /**
     * Perform login with email and password.
     */
    fun login() {
        val state = _uiState.value
        Log.i(TAG, "Starting login process")

        if (state.email.isBlank() || state.password.isBlank()) {
            _uiState.value = state.copy(error = "Please enter email and password")
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            try {
                val user = authManager.login(
                    email = state.email,
                    password = state.password,
                    rememberMe = state.rememberMe
                )
                Log.i(TAG, "Login successful")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    isAuthenticated = true,
                    user = user
                )
            } catch (e: AuthError) {
                Log.e(TAG, "Login failed: ${e.message}")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message
                )
            } catch (e: Exception) {
                Log.e(TAG, "Login failed with unexpected error: ${e.message}")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = "An unexpected error occurred"
                )
            }
        }
    }

    /**
     * Perform login with biometrics.
     */
    fun loginWithBiometrics(activity: FragmentActivity) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            try {
                val user = authManager.authenticateWithBiometrics(activity)
                Log.i(TAG, "Biometric login successful")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    isAuthenticated = true,
                    user = user
                )
            } catch (e: AuthError) {
                Log.e(TAG, "Biometric login failed: ${e.message}")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message
                )
            } catch (e: Exception) {
                Log.e(TAG, "Biometric login failed with unexpected error: ${e.message}")
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = "Biometric authentication failed"
                )
            }
        }
    }

    /**
     * Check if user is already authenticated.
     */
    private fun checkAuthentication() {
        val isAuthenticated = authManager.isAuthenticated
        val storedEmail = authManager.storedEmail ?: ""
        val rememberMe = authManager.isRememberMeEnabled
        val canUseBiometrics = authManager.canUseBiometricAuthentication(getApplication())

        _uiState.value = _uiState.value.copy(
            email = storedEmail,
            rememberMe = rememberMe,
            isAuthenticated = isAuthenticated,
            canUseBiometrics = canUseBiometrics
        )
    }

    /**
     * Logout the user.
     */
    fun logout() {
        authManager.logout()
        _uiState.value = LoginUiState()
    }
}
